src/twisted/conch/ssh/transport.py

```diff
@@ -1275,31 +1279,44 @@ def _ssh_KEX_ECDH_INIT(self, packet):
         pubHostKey = self.factory.publicKeys[self.keyAlg]
         privHostKey = self.factory.privateKeys[self.keyAlg]
 
-        # Get the curve instance
-        try:
-            curve = keys._curveTable[b'ecdsa' + self.kexAlg[4:]]
-        except KeyError:
-            raise UnsupportedAlgorithm('unused-key')
-
-        # Generate the private key
-        ecPriv = ec.generate_private_key(curve, default_backend())
-
-        # Get the public key
-        ecPub = ecPriv.public_key()
-        encPub = ecPub.public_numbers().encode_point()
-
-        # Take the provided public key and transform it into
-        # a format for the cryptography module
-        theirECPub = ec.EllipticCurvePublicNumbers.from_encoded_point(
-                        curve, pktPub).public_key(default_backend())
-
-        # We need to convert to hex,
-        # so we can convert to an int
-        # so we can make it a multiple precision int.
-        sharedSecret = MP(
-                       int(
-                        binascii.hexlify(
-                          ecPriv.exchange(ec.ECDH(), theirECPub)), 16))
+        if self.kexAlg.find(b'25519') != -1:
+            # Turn the public key into a public key object
+            theirECPub = nacl.public.PublicKey(pktPub)
+
+            # Always generate new private and public keys
+            ecPriv = nacl.public.PrivateKey.generate()
+            ecPub = ecPriv.public_key
+            encPub = ecPub._public_key
+            sharedSecret = MP(
+                int(binascii.hexlify(
+                  crypto_scalarmult(ecPriv._private_key,
+                                  theirECPub._public_key)), 16))
+        else:
```

Comment:
One other nice thing about making this an optional dependency would be that we could potentially factor out this extremely long function into a couple of smaller cipher-management functions so as to avoid monolithic changes like having to indent this lengthy "else" block even though the lines aren't being touched.